### PR TITLE
Implement granular contract permissions, Verify that max balance is not exceeded during mint

### PR DESCRIPTION
**Pull request by Alexey1100**
Resolves conflicts between master and #6 

#### 1) Adds a `set permissions` function accepting a 4-bit permissions integer, where each bit represents a role:

    1) account transfer restrictions (wallets admin)
    2) transfer rules (transfer rules admin)
    3) mint/burn (assets admin)
    4) granting permissions (contract admin)
    
    examples:
    Int(1) 0001 – access to account transfer restrictions
    Int(3) 0011 – access to account transfer restrictions and transfer rules
    Int(6) 0110 – access to transfer rules and mint/burn
    Int(8) 1000 – access to granting permissions (= contract admin)
    Int(15) 1111 – access to everything (= contract admin)

#### 2) Returns error if receiver `max balance` is exceeded during mint

### CODE REVIEW

**Comment by Alexey1100** [comment]
Previous discussion regarding permission system:
https://github.com/CoMakery/algorand-security-token/pull/6#discussion_r566619072

File: security_token.py [hunk 1]
```diff
@@ -44,30 +46,47 @@ def approval_program():
...
+    permissions = Btoi(Txn.application_args[1])
+    set_permissions = Seq([
         Assert(And(
             is_contract_admin,
-            Txn.application_args.length() == Int(3),
-            Or(new_admin_type == Bytes("contract admin"), new_admin_type == Bytes("transfer admin")),
-            Txn.accounts.length() == Int(1)
+            Txn.application_args.length() == Int(2),
+            Txn.accounts.length() == Int(1),
+            permissions <= Int(15)
```
Comment (aquabu) [inline security_token.py:83]: 👍

File: tests/mint.test.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  token = await shell.cat(`devnet/Primary/algod.token`).stdout
+  clientV2 = new algosdk.Algodv2(token, server, port)
+  let info = await util.deploySecurityToken(clientV2, adminAccount)
+  appId = info.appId
+
+  //opt in
+  await util.optInApp(clientV2, receiverAccount, appId)
+})
+
+// describe('when receiver max balance is not set', async () => {
```
Comment (aquabu) [inline tests/mint.test.js:26]: Probably should remove the commented out describe blocks or add them back in.

**Review by aquabu — APPROVED** [review]

File: tests/mint.test.js (hunk 2 above)
Comment (Alexey1100) [reply tests/mint.test.js:26]: Yep, as I figure out later describe blocks don't work with async code in Jest.
Let me remove that.

**Pull request merged**